### PR TITLE
chore(setup): add support for downloading libdatadog FFI artifacts during build

### PR DESCRIPTION
**Pull request by sanchda**
This adds the capability for setup.py to acquire libdatadog artifacts.  It generalizes the artifact acquisition code used by libddwaf (hopefully in an agreeable way).

This capability is activated in this PR, but aside from downloading an artifact which is unused in the final wheel, it is unused.  The overhead is on the order of 26 megabytes of disk space for tarball contents

Ensuring that the changes are nonbreaking to libddwaf should be guaranteed by our current tests.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -95,36 +113,27 @@ def is_64_bit_python():
     return sys.maxsize > (1 << 32)
 
 
-class CleanLibraries(CleanCommand):
-    @staticmethod
-    def remove_dynamic_library():
-        shutil.rmtree(LIBDDWAF_DOWNLOAD_DIR, True)
+class Library_Download:
```
Comment (P403n1x87) [inline setup.py:116]:
```suggestion
class LibraryDownload:
```

File: setup.py [hunk 2]
```diff
@@ -95,36 +113,27 @@ def is_64_bit_python():
...
-        }
-        SUFFIX = TRANSLATE_SUFFIX[CURRENT_OS]
+    @classmethod
+    def download_artifacts(cls):
+        suffixes = cls.get_suffixes(CURRENT_OS)
 
         # If the directory exists and it is not empty, assume the right files are there.
         # Use `python setup.py clean` to remove it.
-        if os.path.isdir(LIBDDWAF_DOWNLOAD_DIR) and len(os.listdir(LIBDDWAF_DOWNLOAD_DIR)):
+        if os.path.isdir(cls.download_dir) and len(os.listdir(cls.download_dir)):
```
Comment (P403n1x87) [inline setup.py:130]:
```suggestion
        if os.path.isdir(cls.download_dir) and os.listdir(cls.download_dir):
```

File: setup.py [hunk 3]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+            for suffix in suffixes:
+                original_file = os.path.join(arch_dir, "lib", cls.name + suffix)
+                if os.path.exists(original_file):
+                    renamed_file = os.path.join(arch_dir, "lib", "lib" + cls.name + suffix)
+                    os.rename(original_file, renamed_file)
 
             os.remove(filename)
 
+
+class LibDDWaf_Download(Library_Download):
```
Comment (P403n1x87) [inline setup.py:202]:
```suggestion
class LibDDWafDownload(LibraryDownload):
```

File: setup.py [hunk 4]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+        return archive_dir
+
+    @classmethod
+    def get_suffixes(cls, os):
+        TRANSLATE_SUFFIX = {"Windows": (".dll",), "Darwin": (".dylib",), "Linux": (".so",)}
+        return TRANSLATE_SUFFIX[os]
+
+    @classmethod
+    def run(cls):
+        LibDDWaf_Download.download_artifacts()
```
Comment (P403n1x87) [inline setup.py:225]:
```suggestion
        cls.download_artifacts()
```

File: setup.py [hunk 5]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+
+
+class LibDatadog_Download(Library_Download):
```
Comment (P403n1x87) [inline setup.py:228]:
```suggestion
class LibDatadogDownload(LibraryDownload):
```

File: setup.py [hunk 6]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+        if CURRENT_OS == "Linux":
+            return [
+                "ddtrace/internal/datadog/profiling/include",
+                "ddtrace/internal/datadog/profiling/libdatadog/x86_64/include",
+            ]
+        return []
+
+    @classmethod
+    def run(cls):
+        LibDatadog_Download.download_artifacts()
```
Comment (P403n1x87) [inline setup.py:280]:
```suggestion
    @classmethod
    def run(cls):
        cls.download_artifacts()
```
it looks like we might be able to put this method in the base class?

File: setup.py [hunk 7]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+            "Linux": "unknown-linux-gnu",
+        }
+        tar_osname = osnames[os]
+        archive_dir = "lib%s-%s-%s" % (cls.name, arch, tar_osname)
+        return archive_dir
+
+    @classmethod
+    def get_suffixes(cls, os):
+        TRANSLATE_SUFFIX = {"Windows": (), "Darwin": (), "Linux": (".a", ".h")}
+        return TRANSLATE_SUFFIX[os]
```
Comment (P403n1x87) [inline setup.py:256]: nit: if we make `TRANSLATE_SUFFIX` a class attribute, then we could implement this logic in the base class, and let subclasses simply override the `TRANSLATE_SUFFIX` attribute?

File: setup.py [hunk 8]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
+                "ddtrace/internal/datadog/profiling/libdatadog/x86_64/include",
+            ]
+        return []
+
+    @classmethod
+    def run(cls):
+        LibDatadog_Download.download_artifacts()
+
+
+class Library_Downloader(BuildPyCommand):
```
Comment (P403n1x87) [inline setup.py:283]:
```suggestion
class LibraryDownloader(BuildPyCommand):
```

File: setup.py [hunk 9]
```diff
@@ -139,56 +148,157 @@ def download_dynamic_library():
...
     def run(self):
-        CleanLibraries.remove_dynamic_library()
-        LibDDWaf_Download.download_dynamic_library()
+        CleanLibraries.remove_artifacts()
+        LibDatadog_Download.run()
+        LibDDWaf_Download.run()
```
Comment (P403n1x87) [inline setup.py:287]:
```suggestion
        LibDatadogDownload.run()
        LibDDWafDownload.run()
```

File: setup.py [hunk 10]
```diff
@@ -366,7 +476,7 @@ def get_exts_for(name):
     tests_require=["flake8"],
     cmdclass={
         "build_ext": BuildExtCommand,
-        "build_py": LibDDWaf_Download,
+        "build_py": Library_Downloader,
```
Comment (P403n1x87) [inline setup.py:479]:
```suggestion
        "build_py": LibraryDownloader,
```

File: setup.py (hunk 2 above)
Comment (P403n1x87) [reply setup.py]: Ah this was already coded with `len`. The diff was something else. Feel free to ignore.

Comment (sanchda) [reply setup.py]: Made good sense to me.  Plus--touch the line, own the line, or something.

File: setup.py (hunk 7 above)
Comment (sanchda) [reply setup.py]: Good call, taken in [08737cc](https://github.com/DataDog/dd-trace-py/pull/5940/commits/08737ccc1c2490835fe89d4438b8602e2076be2c)

File: setup.py (hunk 6 above)
Comment (sanchda) [reply setup.py]: Indeed, fixed in 08737cc

File: setup.py (hunk 4 above)
Comment (sanchda) [reply setup.py]: Indeed, fixed in 08737cc

**Review by gnufede — APPROVED** [review]

**Comment by gnufede** [comment]
Neat refactor! ❤️

**Review by avara1986 — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Pull request merged**